Add uncertainties.umath as lambdify module

zarstensen opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->
Fixes #28030

#### Brief description of what is fixed or changed

Adds `umath` as a possible module for `lambdify`.
When this string is passed, the standard math package (used primarily for constants such as pi, e, ...), and the uncertainties.umath package is used as the backends for lambdifying an expression.

#### Other comments


#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers. Formerly, `log(-x)` incorrectly gave `-log(x)`.

* physics.units
  * Corrected a semantical error in the conversion between volt and statvolt which
    reported the volt as being larger than the statvolt.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->

* utilities
  * added `umath` and `unumpy` as possible modules value to `lambdify`, which uses a combination of the standard python math package and the `uncertainties.umath/unumpy` package as a backend, when chosen.

<!-- END RELEASE NOTES -->

zarstensen committed:
lambdify: add uncertainties.umath backend

zarstensen committed:
add author to .mailmap file

zarstensen committed:
Add uncertainties to optional dependencies

moorepants commented on sympy/utilities/tests/test_lambdify.py:
```diff
@@ -728,6 +729,48 @@ def test_scipy_sparse_matrix():
+    res = f(uncertainties.ufloat(2.0, 0.1), uncertainties.ufloat(3.0, 0.2), uncertainties.ufloat(4.0, 0.3))
+
+    assert res.nominal_value == 16
+    assert res.std_dev == 1.268857754044952
```
You'll need to test floats with something like `numpy.testing.assert_allclose()`.

zarstensen replied:
All float comparisons should now make use of `numpy.testing.assert_allclose`.

zarstensen committed:
replace direct float comparisons with assert_allclose

moorepants commented on sympy/utilities/lambdify.py:
```diff
@@ -118,6 +126,7 @@
     "math": (MATH, MATH_DEFAULT, MATH_TRANSLATIONS, ("from math import *",)),
     "cmath": (CMATH, CMATH_DEFAULT, CMATH_TRANSLATIONS, ("import cmath; from cmath import *",)),
     "mpmath": (MPMATH, MPMATH_DEFAULT, MPMATH_TRANSLATIONS, ("from mpmath import *",)),
+    "umath": (UMATH, UMATH_DEFAULT, UMATH_TRANSLATIONS, ("from math import*; from uncertainties.umath import *",)),
```
If you want both math and unmath, should you pass both those into the arguments of lambdify? This would unexpectedly import functions from math.

zarstensen replied:
The wildcard import should now be replaced with an import consisting purely of constants from the math package.
This way nothing unexpected is imported from math.

zarstensen committed:
lambdify: add uncertainties.unumpy backend

zarstensen committed:
lambdify: replace wildcard import with required constants in umath module

zarstensen commented:
It was also quite simple to get `unumpy` to work.
The only "issue" is that any call to a function on a scalar value returns a 0 dimensional array containing the result of the operation, in contrast with numpy which only returns a scalar value if the input is also a scalar.
According to the source code, it seems like this is intentional:
```
  # ! The newly defined functions (uncertainties.unumpy.cos, etc.)
  # do not behave exactly like their NumPy equivalent (numpy.cos,
  # etc.): cos(0) gives an array() and not a
  # numpy.float... (equality tests succeed, though).
```
[_comment location_](https://github.com/lmfit/uncertainties/blob/fead532a300b295011d5ec33d022211818390888/uncertainties/unumpy/core.py#L696C1-L699C59)

Because of this, i decided to keep the `umath` option as well, as it does not have this quirk.
I can remove `umath`, if you think its too much to have both a `umath` and `unumpy` module option.

moorepants commented:
We can support the scalar and array versions. I think that is fine.

zarstensen committed:
lambdify: fix unumpy module sometimes returning 0d array instead of scalar

zarstensen committed:
lambdify: simplify 0d array fix

zarstensen committed:
lamdify: UnumPyPrinter now uses a naming convention consistent with the other printers

zarstensen committed:
Merge branch 'master' into uncertainties-lambdify

zarstensen commented:
Hi @moorepants, are there any other changes i should make or does the PR look good now?

zarstensen committed:
reword UnumpyPrinter docstring and add comment

zarstensen committed:
add author alias

zarstensen committed:
Revert "reword UnumpyPrinter docstring and add comment"

This reverts commit 29dd5436f0d0f8a77e6c26e019457dcbffee9ed0.

Revert "lamdify: UnumPyPrinter now uses a naming convention consistent with the other printers"

This reverts commit 889c417af5f97d8e060c5a64e798e3d5554a972f.

Revert "lambdify: simplify 0d array fix"

This reverts commit 55078ef5aecb4dd353a3326bbee477b012897cf2.

Revert "lambdify: fix unumpy module sometimes returning 0d array instead of scalar"

This reverts commit c4bffebe2f2820acf07b6e20c12651563f18df91.

zarstensen committed:
update tests after 0d fix revert

zarstensen committed:
Replace typo 'umpmath' with 'umath'

zarstensen committed:
Add uncertainties examples to lambdify docstring.

zarstensen commented on sympy/utilities/lambdify.py:
```diff
@@ -329,9 +344,9 @@ def lambdify(args, expr, modules=None, printer=None, use_imps=True,
-        - The strings ``"math"``, ``"cmath"``, ``"mpmath"``, ``"numpy"``,
-          ``"numexpr"``, ``"scipy"``, ``"sympy"``, ``"tensorflow"``,
-          ``"torch"`` or ``"jax"``. This uses the corresponding printer
+        - The strings ``"math"``, ``"cmath"``, ``"mpmath"``, ``"umpmath"``,
```
that is just a silly typo, it should be umath.
i fixed this in [401462f](https://github.com/sympy/sympy/pull/28032/commits/401462f1d8a4488fb0f41dc9d2fb5718350479e5).

zarstensen commented on sympy/utilities/lambdify.py:
```diff
@@ -329,9 +344,9 @@ def lambdify(args, expr, modules=None, printer=None, use_imps=True,
-          ``"torch"`` or ``"jax"``. This uses the corresponding printer
+        - The strings ``"math"``, ``"cmath"``, ``"mpmath"``, ``"umpmath"``,
+          ``"numpy"``, ``"unumpy"``, ``"numexpr"``, ``"scipy"``, ``"sympy"``,
+          ``"tensorflow"``, ``"torch"`` or ``"jax"``. This uses the corresponding printer
```
i have added some examples for umath and unumpy in [7df6999](https://github.com/sympy/sympy/pull/28032/commits/7df699977ac7e7f74ac570f42dbce3d6ef5bd5db).

zarstensen commented:
Finally had some time to look at this, sorry for the long wait.

I believe i have addressed all of your comments in the latest commits, let me know if you want anything else changed.

moorepants commented:
There is an unrelated test failure:

```
==================================== ERRORS ====================================
____________ ERROR collecting sympy/printing/tests/test_llvmjit.py _____________
sympy/printing/tests/test_llvmjit.py:7: in <module>
    import sympy.printing.llvmjitcode as g
sympy/printing/llvmjitcode.py:19: in <module>
    llvm.initialize()
/opt/hostedtoolcache/Python/3.13.7/x64/lib/python3.13/site-packages/llvmlite/binding/initfini.py:17: in initialize
    raise RuntimeError(
Traceback (most recent call last):
  File "/home/runner/work/sympy/sympy/bin/test_optional_dependencies.py", line 150, in <module>
Error:     raise TestsFailedError('Tests and doctests failed')
TestsFailedError: Tests and doctests failed
E   RuntimeError: llvmlite.binding.initialize() is deprecated and will be removed. LLVM initialization is now handled automatically. Please remove calls to this function from your code and check for other behavioral changes that may have occurred due to LLVM updates.
                                DO *NOT* COMMIT!                                
=========================== short test summary info ============================
ERROR sympy/printing/tests/test_llvmjit.py - RuntimeError: llvmlite.binding.initialize() is deprecated and will be removed. LLVM initialization is now handled automatically. Please remove calls to this function from your code and check for other behavioral changes that may have occurred due to LLVM updates.
!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!
======================= 34 deselected, 1 error in 58.66s =======================
Error: Process completed with exit code 1.
```

Maybe you should merge the tip of the master branch here to see if something is fixed on master since this PR has been opened.

moorepants commented:
Actually this llvmlite code seems to be broken in master: #28471
